### PR TITLE
[TK-08010 TK-08011] tx2 proxy

### PR DESCRIPTION
**Pull request by neonphog**
DEPENDS ON https://github.com/holochain/holochain/pull/705

[Video Walkthrough: Tx2 Proxy (8:39)](https://www.youtube.com/watch?v=hy0JJAdnMr8)

- tx2 proxy wrapper for any tx2 frontend
- benchmark for tx2 proxy wrapper over mem transport

### CODE REVIEW

File: crates/kitsune_p2p/proxy/benches/thru.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+fn thru(t: &Arc<Share<Option<Test>>>) {
+    let _g = RUNTIME.enter();
+    RUNTIME.block_on(async {
+        test(t).await;
+    });
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+    let t = RUNTIME.block_on(async { Arc::new(Share::new(Some(Test::new().await))) });
```
Comment (freesig) [inline crates/kitsune_p2p/proxy/benches/thru.rs:206]: Not a huge deal but `criterion_benchmark` is only ever called once so you don't really need the Lazy and you could just create the runtime in this function.

File: crates/kitsune_p2p/proxy/src/tx2.rs [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+struct ProxyConHnd {
+    uniq: Uniq,
+    sub_con: ConHnd,
+    local_digest: CertDigest,
+    remote_digest: CertDigest,
+}
+
+impl std::fmt::Debug for ProxyConHnd {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("ConHnd").finish()
```
Comment (freesig) [inline crates/kitsune_p2p/proxy/src/tx2.rs:35]: Might be useful to include the uniq here for debugging?

File: crates/kitsune_p2p/proxy/src/tx2.rs [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+
+    fn is_closed(&self) -> bool {
+        self.sub_con.is_closed()
+    }
+
+    fn close(&self, _code: u32, _reason: &str) -> BoxFuture<'static, ()> {
+        // TODO - FIXME
+        // we don't want to close the underlying sub_con,
+        // it could be shared for proxying...
+        // do we want to do *anything*?
```
Comment (freesig) [inline crates/kitsune_p2p/proxy/src/tx2.rs:72]: How do we close down the proxy when we are actually finished though?

File: crates/kitsune_p2p/proxy/src/tx2.rs [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+        promote_addr(&remote_addr, &self.remote_digest)
+    }
+
+    fn write(
+        &self,
+        msg_id: MsgId,
+        mut data: PoolBuf,
+        timeout: KitsuneTimeout,
+    ) -> BoxFuture<'static, KitsuneResult<()>> {
+        data.reserve_front(1 + 32 + 32);
```
Comment (freesig) [inline crates/kitsune_p2p/proxy/src/tx2.rs:87]: Might be nice to have these as const

File: crates/kitsune_p2p/proxy/src/tx2.rs [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+                        } else {
+                            let dest = hnd.inner.share_mut(|i, _| {
+                                Ok(i.in_digest_to_sub_con.get(&dest_digest).cloned())
+                            });
+                            match dest {
+                                Ok(Some(d_sub_con)) => {
+                                    let t = KitsuneTimeout::from_millis(1000 * 30);
+                                    // TODO - correct to await here / backpressure?
+                                    // TODO - respond with errors?
+                                    let _ = d_sub_con.write(msg_id, data, t).await;
```
Comment (freesig) [inline crates/kitsune_p2p/proxy/src/tx2.rs:361]: Does this make use single threaded for all proxy messages?

File: crates/kitsune_p2p/proxy/src/tx2.rs [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+                                }
+                                // TODO - FIXME
+                                e => panic!("{:?}", e),
```
Comment (freesig) [inline crates/kitsune_p2p/proxy/src/tx2.rs:364]: I'm guessing merging this PR won't make it be used in production but we'd obviously have to fix this before then :)

**Review by freesig — APPROVED** [review]
Looks like a good start on the proxy. A few places we'd won't to remove panics but I imagine this PR will be built on before being used in production.

File: crates/kitsune_p2p/proxy/src/tx2.rs (hunk 3 above)
Comment (neonphog) [reply crates/kitsune_p2p/proxy/src/tx2.rs:72]: The close at the endpoint level will close everything properly... there just isn't yet a way to close down individual connections.

We probably want to fix this somehow at some point, but I don't think it's dire. Will leave the `TODO - FIXME` in there.

File: crates/kitsune_p2p/proxy/benches/thru.rs [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    t.test().await;
+    this.share_mut(move |i, _| {
+        *i = Some(t);
+        Ok(())
+    })
+    .unwrap();
+}
+
+fn thru(t: &Arc<Share<Option<Test>>>) {
+    let _g = RUNTIME.enter();
```
[the remark this answers is not in the document]
Comment (neonphog) [reply crates/kitsune_p2p/proxy/benches/thru.rs]: fixed.

File: crates/kitsune_p2p/proxy/benches/thru.rs (hunk 1 above)
Comment (neonphog) [reply crates/kitsune_p2p/proxy/benches/thru.rs]: fixed.

File: crates/kitsune_p2p/proxy/src/tx2.rs (hunk 2 above)
Comment (neonphog) [reply crates/kitsune_p2p/proxy/src/tx2.rs]: Good idea, done.

File: crates/kitsune_p2p/proxy/src/tx2.rs (hunk 4 above)
Comment (neonphog) [reply crates/kitsune_p2p/proxy/src/tx2.rs]: fixed.

File: crates/kitsune_p2p/proxy/src/tx2.rs (hunk 5 above)
Comment (neonphog) [reply crates/kitsune_p2p/proxy/src/tx2.rs]: Yes. Interestingly enough, performance seems to degrade slightly when I run this logic with max cpu_count semaphore tasks... 

But there is the possibility of backed up writes slowing us down, and we've got 3 channels to work with, so I switched this to use `for_each_concurrent(3, ...` which doesn't seem to have any effect on performance for our benchmark, but could theoretically speed up that back-pressured writer case?

File: crates/kitsune_p2p/proxy/src/tx2.rs (hunk 6 above)
Comment (neonphog) [reply crates/kitsune_p2p/proxy/src/tx2.rs]: Haha, yeah, missed these before making the PR.

Removed all the panics. In a couple cases I'm just logging out the errors where we should be responding to the requestor with an error, but we can add that in a future iteration.

**Pull request merged**